src/drivers/kinetis/uart.c

```diff
@@ -0,0 +1,269 @@
+
+    // make isr available
+    which_uart = 0;
+    isr_active = false;
+
+    //Enable transmitter and receiver of UART (and interrupts)
+    if(conf.configure_interrupts > 0) {
+        asm("cpsid i");
+        // set RWFIFO[RXWATER] such that interrupts are actually generated
+        if(conf.rwfifo_sz > 0) {
+            conf.uart_base->RWFIFO = conf.rwfifo_sz;
+            conf.uart_base->PFIFO |= (UART_PFIFO_RXFE_MASK);
+            conf.uart_base->CFIFO |=
+                (UART_CFIFO_TXFLUSH_MASK | UART_CFIFO_RXFLUSH_MASK);
+        }
+        else {
+            conf.uart_base->RWFIFO = 1;
+        }
+
+        // set TWFIFO[TXWATER] such that interrupts are actually generated
+        if(conf.twfifo_sz > 0) {
+            conf.uart_base->TWFIFO = conf.twfifo_sz;
+            conf.uart_base->PFIFO |= (UART_PFIFO_TXFE_MASK);
+            conf.uart_base->CFIFO |=
+                (UART_CFIFO_TXFLUSH_MASK | UART_CFIFO_RXFLUSH_MASK);
+        }
+        // no else, setting twfifo is not required to get TDRE int
+
+        // turn on rx, tx, rx interrupt, tx interrupt
+        conf.uart_base->C2 |= UART_C2_RE_MASK | UART_C2_TE_MASK
+            | UART_C2_RIE_MASK | UART_C2_TIE_MASK;
+
+        NVIC_EnableIRQ(conf.irqn);
+        NVIC->IP[conf.irqn] = conf.priority;
+        // generate DMA requests instead of interrupts
+        /*conf.uart_base->C5 |= UART_C5_TDMAS_MASK;*/
+        // enable tx fifo
+        conf.uart_base->PFIFO |= UART_PFIFO_TXFE_MASK;
+        asm("cpsie i");
+    }
+    else {
+        conf.uart_base->C2 |= UART_C2_RE_MASK | UART_C2_TE_MASK;
+    }
+
+    return ftab_open((ftab_entry_t){
+            .context = (void*)&contexts[which],
+            .write = uart_write,
+            .read = uart_read,
+            .close = uart_close,
+        }
+    );
+}
+
+unsigned int uart_write(uart_context *context, char *buf, unsigned int bytes) {
+    unsigned int bytes_written = 0;
+    if(bytes == 0) {
+        bytes_written = context->txq->cap - context->txq->size;
+        goto done;
+    }
+    do {
+        queue_push(context->txq, buf[bytes_written]);
+        bytes_written += (context->txq->op_ok) ? 1:0;
+    } while(context->txq->op_ok && bytes_written < bytes);
+    if(bytes_written > 0) context->uart_base->C2 |= UART_C2_TIE_MASK;
+done:
+    return bytes_written;
+}
+
+
+unsigned int uart_read(uart_context *context, char *buf, unsigned int bytes) {
+    unsigned int bytes_read = 0;
+    if(bytes == 0) {
+        bytes_read = context->rxq->size;
+        goto done;
+    }
+    do {
+        char c = queue_pop(context->rxq);
+        if(context->rxq->op_ok) {
+            buf[bytes_read] = c;
+        }
+        bytes_read += (context->rxq->op_ok) ? 1:0;
+    } while(context->rxq->op_ok && bytes_read < bytes);
+done:
+    return bytes_read;
+}
+
+
+unsigned int uart_close(uart_context *context) { return 0; }
+
+
+void uart_isr(int which_uart){
+    uint8_t sz = 0;
+    asm("cpsid i");

```

Comment:
whats the purpose of turning off interrupts here?

Reply:
No interruption during the ISR - if we allow interruption, the data may be lost (this was likely one of the causes of bugs in the asm uart driver)